File: src/bin/morelogic/LStringStatement.java
package bin.morelogic;

import arc.graphics.Color;
import arc.scene.ui.layout.Table;
import mindustry.graphics.Pal;
import mindustry.logic.*;
import mindustry.logic.LExecutor.LInstruction;
import mindustry.logic.LExecutor.Var;

/**
 * @author bin
 */
@SuppressWarnings({"AlibabaUndefineMagicConstant", "AlibabaClassNamingShouldBeCamel"})
public class LStringStatement {

    public static class AppendStringStatement extends LStatement {
        public static final String ID = "bin_AppendString";
        public String dest = "result", a = "a", b = "b";

        public static AppendStringStatement read(String[] tokens) {
            final var result = new AppendStringStatement();
            final var length = tokens.length;
            if (length > 1) {
                result.dest = tokens[1];
            }
            if (length > 2) {
                result.a = tokens[2];
            }
            if (length > 3) {
                result.b = tokens[3];
            }
            return result;
        }

        @Override
        public void build(Table table) {
            table.clearChildren();
            field(table, dest, str -> dest = str);
            table.add(" = ");
            row(table);
            field(table, a, str -> a = str);
            table.add(" append ");
            field(table, b, str -> b = str);
        }

        @Override
        public LInstruction build(LAssembler builder) {
            return new AppendInstruction(builder.var(a), builder.var(b), builder.var(dest));
        }

        @Override
        public Color color() {
            return Pal.logicOperations;
        }

        @Override
        public void write(final StringBuilder builder) {
            builder.append(ID);
            append(builder,
                    dest,
                    a,
                    b
            );
        }

        private static class AppendInstruction implements LInstruction {
            public int a, b, dest;

            public AppendInstruction(int a, int b, int dest) {
                this.a = a;
                this.b = b;
                this.dest = dest;
            }

            @Override
            public void run(LExecutor exec) {
                String v = str(exec.var(a)), v2 = str(exec.var(b));
                exec.setobj(dest, v + v2);
            }
        }

    }

    public static String str(Var v) {
        if (v.isobj) {
            final var objval = v.objval;
            return objval instanceof String ? (String) objval : String.valueOf(objval);
        } else {
            return Double.toString(v.numval);
        }
    }

    public static void append(StringBuilder builder, String... strings) {
        for (final String s : strings) {
            builder.append(" ").append(s);
        }
    }
}

File: src/bin/morelogic/LStringStatement.kt
package bin.morelogic

import arc.scene.ui.Button
import arc.scene.ui.layout.Table
import bin.morelogic.statement.StringOp
import mindustry.logic.*
import mindustry.logic.LExecutor.LInstruction
import mindustry.ui.Styles

/**
 * TODO: 字符串的切片，插入，拆除，判断
 * @author bin
 */
object LStringStatement {

	@JvmStatic
	fun StringBuilder.appendBySpace(vararg strings: String) {
		for (s in strings) {
			append(" ").append(s)
		}
	}

	fun LAssembler.vars(vararg symbols: String): IntArray {
		return IntArray(symbols.size) {
			`var`(symbols[it])
		}
	}

	/**
	 * 字符串拼接
	 */
	class AppendStringStatement : LStatement() {
		private var op = StringOp.Length
		private var dest = "ab"
		private var a = "a"
		private var b = "b"
		private var c = "c"
		override fun build(table: Table) = rebuild(table)

		private fun rebuild(table: Table) {
			table.clearChildren()
			field(table, dest) { dest = it }
			table.add(" = ")
			if (op.args == 1) {
				opButton(table, table)
				field(table, a) { a = it }
				return
			}
			row(table)
			if (LCanvas.useRows()) {
				table.left()
				table.row()
				table.table { c: Table ->
					c.color.set(category().color)
					c.left()
					funcs(c, table)
				}.colspan(2).left()
				return
			}
			funcs(table, table)
		}

		private fun funcs(table: Table, parent: Table) {
			opButton(table, parent)
			field(table, a) { a = it }
			field(table, b) { b = it }
		}

		private fun opButton(table: Table, parent: Table) {
			table.button({ b: Button ->
				b.label { op.symbol }
				b.clicked {
					showSelect(b, StringOp.all, op) { o: StringOp ->
						op = o
						rebuild(parent)
					}
				}
			}, Styles.logict) {}.size(64f, 40f).pad(4f).color(table.color)
		}

		override fun build(builder: LAssembler): LInstruction {
			return op(builder.vars(dest, a, b))
		}

		override fun category(): LCategory = LCategory.operation

		override fun write(builder: StringBuilder) {
			builder.append(ID)
			builder.appendBySpace(op.name, dest, a, b)
		}

		companion object {
			const val ID = "bin_String"
			fun read(tokens: Array<String>): LStatement {
				return AppendStringStatement().apply {
					val length = tokens.size
					var i = 0
					if (length > 0) op = StringOp.value(tokens[i++])
					if (length > i) dest = tokens[i++]
					if (length > i) a = tokens[i++]
					if (length > i) b = tokens[i]
				}
			}
		}
	}
}
